netplan/cli/sriov.py

```diff
@@ -0,0 +1,278 @@
+            # created later - but store, for convenience, all the valid
+            # VFs that we encounter so far
+            vfs[ethernet] = None
+
+
+def set_numvfs_for_pf(pf, vf_count):
+    devdir = os.path.join('/sys/class/net', pf, 'device')
+    numvfs_path = os.path.join(devdir, 'sriov_numvfs')
+    totalvfs_path = os.path.join(devdir, 'sriov_totalvfs')
+    try:
+        with open(numvfs_path) as f:
+            vf_current = int(f.read().strip())
+        with open(totalvfs_path) as f:
+            vf_max = int(f.read().strip())
+    except IOError as e:
+        raise RuntimeError('failed parsing sriov_numvfs/sriov_totalvfs for %s: %s' % (pf, str(e)))
+    except ValueError:
+        raise RuntimeError('invalid sriov_numvfs/sriov_totalvfs value for %s' % pf)
+
+    if vf_count > vf_max:
+        raise ConfigurationError(
+            'cannot allocate more VFs for PF %s than supported: %s > %s (sriov_totalvfs)' % (pf, vf_count, vf_max))
+
+    if vf_count <= vf_current:
+        # XXX: this might be a wrong assumption, but I assume that
+        #  the operation of adding/removing VFs is very invasive,
+        #  so it makes no sense to decrease the number of VFs if
+        #  less are needed - leaving the unused ones unconfigured?
+        logging.debug('the %s PF already defines more VFs than required (%s > %s), skipping' % (pf, vf_current, vf_count))
+        return False
+
+    try:
+        with open(numvfs_path, 'w') as f:
+            f.write(str(vf_count))
+    except IOError as e:
+        bail = True
+        if e.errno == 16:  # device or resource busy
+            logging.warning('device or resource busy while setting sriov_numvfs for %s, trying workaround' % pf)
+            try:
+                # doing this in two open/close sequences so that
+                # it's as close to writing via shell as possible
+                with open(numvfs_path, 'w') as f:
+                    f.write('0')
+                with open(numvfs_path, 'w') as f:
+                    f.write(str(vf_count))
+            except IOError as e_inner:
+                e = e_inner
+            else:
+                bail = False
+        if bail:
+            raise RuntimeError('failed setting sriov_numvfs to %s for %s: %s' % (vf_count, pf, str(e)))
+
+    return True
+
+
+def perform_hardware_specific_quirks(pf):
+    devdir = os.path.join('/sys/class/net', pf, 'device')
+    try:
+        with open(os.path.join(devdir, 'vendor')) as f:
+            device_id = f.read().strip()[2:]
+        with open(os.path.join(devdir, 'device')) as f:
+            vendor_id = f.read().strip()[2:]
+    except IOError as e:
+        raise RuntimeError('could not determine vendor and device ID of %s: %s' % (pf, str(e)))
+
+    combined_id = ':'.join([vendor_id, device_id])
+    quirk_devices = ()  # TODO: add entries to the list
+    if combined_id in quirk_devices:
+        # some devices need special handling, so this is the place
```

Comment:
So I didn't actually look for it too much, but I added this as a precaution as Stephane Graber and a few other people, who had a lot of experience with SR-IOV devices, said that many Intel and Mellanox ones required a kernel module reload for the actual new sriov_numvfs value to get accepted. That being said, on the SR-IOV Intel card that I have been given access to everything works as expected (no quirks needed). Same for documentation - both the Mellanox and Intel documentation on SR-IOV handling on Linux do not mention anything about any issues like that. I assume this might be the case for some cheaper cards perhaps?
Anyway, I will try getting some more info on that. I'll at least try to explain the situation there better in comments.